src/datanode/src/region_server.rs

```diff
@@ -112,13 +161,150 @@ enum RegionChange {
+                .await
+                .with_context(|_| GetRegionMetadataSnafu {
+                    region: engine.name(),
+                    region_id,
+                })?;
+        let table_provider = DummyTableProvider {
+            region_id,
+            engine,
+            metadata,
+            scan_request: Default::default(),
+        };
+        let schema_provider = DummySchemaProvider {
+            table: table_provider,
+        };
+        let catalog_provider = DummyCatalogProvider {
+            schema: schema_provider,
+        };
+        let catalog_list = Self {
+            catalog: catalog_provider,
+        };
+        Ok(catalog_list)
+    }
+}
+
+impl CatalogList for DummyCatalogList {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn register_catalog(
+        &self,
+        _name: String,
+        _catalog: Arc<dyn CatalogProvider>,
+    ) -> Option<Arc<dyn CatalogProvider>> {
+        None
+    }
+
+    fn catalog_names(&self) -> Vec<String> {
+        vec![]
+    }
+
+    fn catalog(&self, _name: &str) -> Option<Arc<dyn CatalogProvider>> {
+        Some(Arc::new(self.catalog.clone()))
+    }
+}
+
+/// For [DummyCatalogList].
+#[derive(Clone)]
+struct DummyCatalogProvider {
+    schema: DummySchemaProvider,
+}
+
+impl CatalogProvider for DummyCatalogProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema_names(&self) -> Vec<String> {
+        vec![]
+    }
+
+    fn schema(&self, _name: &str) -> Option<Arc<dyn SchemaProvider>> {
+        Some(Arc::new(self.schema.clone()))
+    }
+}
+
+/// For [DummyCatalogList].
+#[derive(Clone)]
+struct DummySchemaProvider {
+    table: DummyTableProvider,
+}
+
+#[async_trait]
+impl SchemaProvider for DummySchemaProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn table_names(&self) -> Vec<String> {
+        vec![]
+    }
+
+    async fn table(&self, _name: &str) -> Option<Arc<dyn TableProvider>> {
+        Some(Arc::new(self.table.clone()))
+    }
+
+    fn table_exist(&self, _name: &str) -> bool {
+        true
+    }
+}
+
+/// For [TableProvider](datafusion::datasource::TableProvider) and [DummyCatalogList]
+#[derive(Clone)]
+struct DummyTableProvider {
+    region_id: RegionId,
+    engine: RegionEngineRef,
+    metadata: RegionMetadataRef,
+    scan_request: Arc<Mutex<ScanRequest>>,
+}
+
+#[async_trait]
+impl TableProvider for DummyTableProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.metadata.schema.arrow_schema().clone()
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        _state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[DfExpr],
+        limit: Option<usize>,
+    ) -> DfResult<Arc<dyn DfPhysicalPlan>> {
+        let mut request = self.scan_request.lock().unwrap().clone();
```

Comment:
question: Why do we put a scan request in the provider? Could we create a fresh scan request each time?

Reply:
Keeping a request makes it possible to change the request field before calling `scan`